src/main/java/org/opensearch/knn/search/NativeEngineKnnVectorQuery.java

```diff
@@ -0,0 +1,282 @@
+package org.opensearch.knn.search;
+
+import lombok.RequiredArgsConstructor;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.MatchNoDocsQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.TopDocs;
+import org.apache.lucene.search.TotalHits;
+import org.apache.lucene.search.Weight;
+import org.opensearch.knn.index.query.KNNQuery;
+import org.opensearch.knn.index.query.KNNScorer;
+import org.opensearch.knn.index.query.KNNWeight;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.Callable;
+
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+
+@RequiredArgsConstructor
+public class NativeEngineKnnVectorQuery extends Query {
+
+    private final KNNQuery knnQuery;
+
+    public Query rewrite(final IndexSearcher indexSearcher) throws IOException {
+        final IndexReader reader = indexSearcher.getIndexReader();
+        final KNNWeight knnWeight = (KNNWeight) knnQuery.createWeight(indexSearcher, ScoreMode.COMPLETE, 1);
+        List<LeafReaderContext> leafReaderContexts = reader.leaves();
+
+        List<Callable<TopDocs>> tasks = new ArrayList<>(leafReaderContexts.size());
+        for (LeafReaderContext leafReaderContext : leafReaderContexts) {
+            tasks.add(() -> searchLeaf(leafReaderContext, knnWeight));
+        }
+        TopDocs[] perLeafResults = indexSearcher.getTaskExecutor().invokeAll(tasks).toArray(TopDocs[]::new);
+        TopDocs topK = TopDocs.merge(knnQuery.getK(), perLeafResults);
+        if (topK.scoreDocs.length == 0) {
+            return new MatchNoDocsQuery();
+        }
+        return createRewrittenQuery(reader, topK);
+    }
+
+    private Query createRewrittenQuery(IndexReader reader, TopDocs topK) {
+        int len = topK.scoreDocs.length;
+        Arrays.sort(topK.scoreDocs, Comparator.comparingInt(a -> a.doc));
```

Comment:
why we are comparing docId here? shouldn't this be score.

Reply:
So the DocAndScoreQuery expects docs in sorted order. DocAndScoreQuery uses this sorted docIds to divide the results per leaf. This leverages concurrent execution to then put docs in TopDocsCollector faster. 

I did not change it because Its an optimization that will help control latencies when k is large and concurrent segment search is on. Alternative is to have a Query which has a scorer that returns all results in the first leaf and then returns empty scorer for subsequent leaves